src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -285,32 +290,39 @@ private async Task CheckBatchContinuouslyAsync()
         {
             this.logger.LogTrace("()");
 
-            int cumulativeTicks = 0;
-            bool saveBatch = false;
-
             while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
             {
+                bool saveBatch = false;
+
                 try
                 {
-                    await Task.Delay(1000);
-
-                    cumulativeTicks++;
-
-                    if (cumulativeTicks >= BatchMaxSaveIntervalSeconds)
+                    using (CancellationTokenSource timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.nodeLifetime.ApplicationStopping))
                     {
-                        saveBatch = true;
-                        cumulativeTicks = 0;
+                        timeoutTokenSource.CancelAfter(BatchMaxSaveIntervalSeconds * 1000);
```

Comment:
Why use cancellation token source instead of timer? Either use timer or a separated task which completes after certain time

Reply:
The previous implementation used `Task.Delay` as a timer anyway, which you wanted changed when I used it in the new implementation. I don't see using cancellation as a timer like this as being particularly problematic, we do the same thing in the P2P code, for example.

Reply:
that feels very heavy. Task.Delay is quite slim compared to that. in P2P there is a onetime thing, this is repeated timer, so it feels wrong